libflux/src/lib.rs

```diff
@@ -9,24 +9,58 @@ pub mod scanner;
 pub mod semantic;
 
 use std::ffi::*;
-use std::os::raw::c_char;
+use std::error::Error;
+use std::os::raw::{c_void, c_char};
 
 use parser::Parser;
 
+pub mod ctypes {
+    include!(concat!(env!("OUT_DIR"), "/ctypes.rs"));
+}
+use ctypes::*;
+
+struct ErrorHandle {
+    err: Box<dyn Error>,
+}
+
 #[no_mangle]
-pub extern "C" fn flux_parse(cstr: *mut c_char) -> *mut ast::File {
+pub extern "C" fn flux_parse(cstr: *mut c_char) -> *mut flux_ast_t {
     let buf = unsafe {
         CStr::from_ptr(cstr).to_bytes()
     };
     let s = String::from_utf8(buf.to_vec()).unwrap();
     let mut p = Parser::new(&s);
     let file = p.parse_file(String::from(""));
-    return Box::into_raw(Box::new(file));
+    return Box::into_raw(Box::new(file)) as *mut flux_ast_t;
+}
+
+#[no_mangle]
+pub extern "C" fn flux_ast_marshal_json(ast: *mut flux_ast_t, buf: *mut flux_buffer_t) -> *mut flux_error_t {
+    let self_ = unsafe { &*(ast as *mut ast::File) } as &ast::File;
+    let data = match serde_json::to_vec(self_) {
+        Ok(v) => v,
+        Err(err) => {
+            let errh = ErrorHandle{ err: Box::new(err) };
+            return Box::into_raw(Box::new(errh)) as *mut flux_error_t
+        },
+    };
+
+    let buffer = unsafe { &mut *buf };
+    buffer.len = data.len();
+    buffer.data = Box::into_raw(data.into_boxed_slice()) as *mut c_void;
+    return std::ptr::null_mut();
+}
+
+#[no_mangle]
+pub extern "C" fn flux_error_str(err: *mut flux_error_t) -> *mut c_char {
+    let e = unsafe { &*(err as *mut ErrorHandle) };
+    let s = CString::new(e.err.description()).unwrap();
+    return s.into_raw();
```

Comment:
Why is the unsafe pointer required/preferred and how is it handled? I'd just like to understand more about your thought process here. Still learning how pointers work in Rust.

Reply:
I did see that you mentioned that this leads to better, more-consistent type names. Just wondering if there were additional thoughts that led to this approach.

Reply:
The pointer type doesn't really matter since those types don't really exist anyway. In C, you are allowed to forward declare a type and use it as a pointer without declaring what's inside of it since what's inside of it ends up becoming part of the ABI (application binary interface). That ends up meaning that an exposed struct requires updating the library major version whenever you modify one that is publicly exposed so you should avoid publicly exposing structs that the caller doesn't need. It also prevents anybody from messing with the internals of your struct if you want to handle it in a certain way.

The reason why an unsafe pointer is required is because of Rust's memory management. If we don't call `into_raw()` then Rust will see that the lifetime of the box is finished and it will free the memory associated with it. When we turn it into a raw pointer, we are telling Rust that we are taking ownership of that memory and that we will handle the lifetime. This is needed because we are returning memory from Rust to Go and then using it with an indeterminate lifetime. It's unsafe because the memory guarantees that Rust gives are gone, but we need to do it because we need to control the lifetime of this pointer manually rather than let the Rust runtime take care of it.

The type names are mostly just to make it easier to follow the code and see that it matches the C code. You can have all of these return `void *` and they would work fine since the ABI's would be compatible. Once a box is turned into a raw pointer, it really doesn't matter what's inside of it. That's also the unsafe part because there's no guarantee that `*mut flux_error_t` is actually an error type other than us giving a pinky-promise that we won't do anything bad.

Reply:
Thanks for the detailed explanation. Helped a lot